File: java/javasearch/src/test/java/javasearch/SearchOptionsTest.java
package javasearch;

import org.json.simple.parser.ParseException;
import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.*;

public class SearchOptionsTest {

    public SearchOptionsTest() {

    }

    @Test
    public final void testSettingsFromMinimalArgs() {
        String[] args = new String[]{"-s", "Search", "."};
        try {
            SearchOptions searchOptions = new SearchOptions();
            SearchSettings settings = searchOptions.settingsFromArgs(args);
            assertFalse(settings.getArchivesOnly());
            assertFalse(settings.getDebug());
            assertTrue(settings.getExcludeHidden());
            assertFalse(settings.getFirstMatch());
            assertEquals(0, settings.getLinesAfter());
            assertEquals(0, settings.getLinesBefore());
            assertFalse(settings.getListDirs());
            assertFalse(settings.getListFiles());
            assertFalse(settings.getListLines());
            assertEquals(150, settings.getMaxLineLength());
            assertFalse(settings.getMultiLineSearch());
            assertEquals(1, settings.getPaths().size());
            assertTrue(settings.getPaths().contains("."));
            assertTrue(settings.getPrintResults());
            assertFalse(settings.getPrintUsage());
            assertFalse(settings.getPrintVersion());
            assertFalse(settings.getSearchArchives());
            assertFalse(settings.getUniqueLines());
            assertFalse(settings.getVerbose());
        } catch (SearchException e) {
            System.out.println("SearchException: " + e.getMessage());
            fail();
        } catch (ParseException | IOException e) {
            System.out.println("Exception: " + e.getMessage());
            fail();
        }
    }

    @Test
    public final void testSettingsFromValidArgs() {
        String[] args = new String[]{"-x", "java,scala", "-s", "Search", "."};
        try {
            SearchOptions searchOptions = new SearchOptions();
            SearchSettings settings = searchOptions.settingsFromArgs(args);
            assertEquals(2, settings.getInExtensions().size());
            assertTrue(settings.getInExtensions().contains("java"));
            assertTrue(settings.getInExtensions().contains("scala"));
            assertEquals(1, settings.getSearchPatterns().size());
            assertEquals("Search", settings.getSearchPatterns().toArray()[0].toString());
        } catch (SearchException e) {
            System.out.println("SearchException: " + e.getMessage());
            fail();
        } catch (ParseException | IOException e) {
            System.out.println("Exception: " + e.getMessage());
            fail();
        }
    }

    @Test
    public final void testSettingsFromJson() {
        StringBuilder json = new StringBuilder("{\n")
                .append("  \"path\": \"~/src/xsearch/\",\n")
                .append("  \"in-ext\": [\"js\",\"ts\"],\n")
                .append("  \"out-dirpattern\": \"node_module\",\n")
                .append("  \"out-filepattern\": [\"temp\"],\n")
                .append("  \"searchpattern\": \"Searcher\",\n")
                .append("  \"linesbefore\": 2,\n")
                .append("  \"linesafter\": 2,\n")
                .append("  \"debug\": true,\n")
                .append("  \"allmatches\": false,\n")
                .append("  \"includehidden\": false,\n")
                .append("}");
        try {
            SearchOptions searchOptions = new SearchOptions();
            SearchSettings settings = new SearchSettings();
            searchOptions.settingsFromJson(json.toString(), settings);

            assertEquals(1, settings.getPaths().size());
            assertTrue(settings.getPaths().contains("~/src/xsearch/"));

            assertEquals(2, settings.getInExtensions().size());
            assertTrue(settings.getInExtensions().contains("js"));
            assertTrue(settings.getInExtensions().contains("ts"));

            assertEquals(1, settings.getOutDirPatterns().size());
            assertEquals("node_module", settings.getOutDirPatterns().toArray()[0].toString());

            assertEquals(1, settings.getOutFilePatterns().size());
            assertEquals("temp", settings.getOutFilePatterns().toArray()[0].toString());

            assertEquals(1, settings.getSearchPatterns().size());
            assertEquals("Searcher", settings.getSearchPatterns().toArray()[0].toString());

            assertEquals(2, settings.getLinesBefore());
            assertEquals(2, settings.getLinesAfter());

            assertTrue(settings.getDebug());
            assertTrue(settings.getFirstMatch());
            assertTrue(settings.getExcludeHidden());
        } catch (ParseException e) {
            System.out.println("ParseException: " + e.getMessage());
            fail();
        } catch (IOException e) {
            System.out.println("IOException: " + e.getMessage());
            fail();
        }
    }
}

File: kotlin/ktsearch/src/test/kotlin/ktsearch/SearchOptionsTest.kt
package ktsearch

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * @author cary on 7/30/16.
 */
class SearchOptionsTest {

    @Test
    fun testSettingsFromMinimalArgs() {
        val args = arrayOf("-s", "Search", ".")
        val searchOptions = SearchOptions()
        val settings = searchOptions.settingsFromArgs(args)
        assertFalse(settings.archivesOnly)
        assertFalse(settings.debug)
        assertFalse(settings.firstMatch)
        assertFalse(settings.includeHidden)
        assertEquals(0, settings.linesAfter)
        assertEquals(0, settings.linesBefore)
        assertFalse(settings.printDirs)
        assertFalse(settings.printFiles)
        assertFalse(settings.printLines)
        assertEquals(150, settings.maxLineLength)
        assertFalse(settings.multiLineSearch)
        assertEquals(1, settings.paths.size)
        assertEquals(".", settings.paths.first())
        assertTrue(settings.printResults)
        assertFalse(settings.printUsage)
        assertFalse(settings.printVersion)
        assertFalse(settings.searchArchives)
        assertFalse(settings.uniqueLines)
        assertFalse(settings.verbose)
    }

    @Test
    fun testSettingsFromValidArgs() {
        val args = arrayOf("-x", "java,scala", "-s", "Search", ".")
        val searchOptions = SearchOptions()
        val settings = searchOptions.settingsFromArgs(args)
        assertEquals(2, settings.inExtensions.size)
        assertTrue(settings.inExtensions.contains("java"))
        assertTrue(settings.inExtensions.contains("scala"))
        assertEquals(1, settings.searchPatterns.size)
        assertEquals("Search", settings.searchPatterns.first().toString())
        assertEquals(1, settings.paths.size)
        assertEquals(".", settings.paths.first())
    }

    @Test
    fun testSettingsFromJson() {
        val json = """{
                 |  "path": "~/src/xsearch/",
                 |  "in-ext": ["js","ts"],
                 |  "out-dirpattern": ["build", "node_module", "tests", "typings"],
                 |  "out-filepattern": ["gulpfile", "\\.min\\."],
                 |  "searchpattern": "Searcher",
                 |  "linesbefore": 2,
                 |  "linesafter": 2,
                 |  "debug": true,
                 |  "allmatches": false,
                 |  "includehidden": false
                 |}""".trimMargin()
        val searchOptions = SearchOptions()
        val settings = searchOptions.settingsFromJson(json, getDefaultSettings())

        assertEquals(1, settings.paths.size)
        assertEquals("~/src/xsearch/", settings.paths.first())

        assertEquals(2, settings.inExtensions.size)
        assertTrue(settings.inExtensions.contains("js"))
        assertTrue(settings.inExtensions.contains("ts"))

        assertEquals(4, settings.outDirPatterns.size)
        assertEquals(1, settings.outDirPatterns.count {it.pattern == "node_module"})

        assertEquals(2, settings.outFilePatterns.size)
        assertEquals(1, settings.outFilePatterns.count {it.pattern == "gulpfile"})

        assertEquals(1, settings.searchPatterns.size)
        assertEquals("Searcher", settings.searchPatterns.first().pattern)

        assertEquals(2, settings.linesBefore)
        assertEquals(2, settings.linesAfter)

        assertTrue(settings.debug)
        assertTrue(settings.firstMatch)
        assertFalse(settings.includeHidden)
    }
}
